README.md
# Gauss - Dantzig Selector: Aggregation over random models (GDS-ARM)

<img src="GDSARM_logo_v1.jfif" width="350px" align="right" style="padding-left:10px;background-color:white;" />

#### *Rakhi Singh, agrakhi@gmail.com*
#### *John Stufken, jstufken@uncg.edu*

<!-- badges: start -->

[![CRAN Status](https://www.r-pkg.org/badges/version/GDSARM)](https://cran.r-project.org/package=GDSARM)
[![R-CMD-check](https://github.com/agrakhi/GDSARM/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/agrakhi/GDSARM/actions/workflows/R-CMD-check.yaml)
[![pkgdown](https://github.com/agrakhi/GDSARM/actions/workflows/pkgdown.yaml/badge.svg)](https://github.com/agrakhi/GDSARM/actions/workflows/pkgdown.yaml)
[![R](https://github.com/agrakhi/GDSARM/actions/workflows/r.yml/badge.svg)](https://github.com/agrakhi/GDSARM/actions/workflows/r.yml)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/0o2rg3h22bbb4yt7/branch/main?svg=true)](https://ci.appveyor.com/project/agrakhi/GDSARM/branch/main)
[![CRAN_Download_Badge](https://cranlogs.r-pkg.org/badges/grand-total/GDSARM)](https://cranlogs.r-pkg.org/badges/grand-total/GDSARM)

<!-- badges: end -->

Screening experiments are useful for screening out a small number of truly impor-
tant factors from a large number of potentially important factors. The Gauss-Dantzig
Selector (GDS) is often the preferred analysis method for screening experiments. Just
considering main-effects models can result in erroneous conclusions, but including in-
teraction terms, even if restricted to two-factor interactions, increases the number of
model terms dramatically and challenges the GDS analysis. We propose a new analysis method, 
called Gauss-Dantzig Selector Aggregation over Random Models (***GDS-ARM***), which performs a GDS analysis on multiple models that include only some
randomly selected interactions. Please see [Singh and Stufken (2022)](https://arxiv.org/abs/2205.13497) for more details.

## Development status

This package is currently in the submission to CRAN. The programs are in stable development phase. Any major changes to include factors at more than two-levels will be added over time.

## Installation

(yet to come in the next week)
Most stable version will be pushed to CRAN and can be installed directly from CRAN in a week's time:

```{r}
install.packages("GDSARM")
```

The latest version of the package under development can be installed from GitHub:

```{r}
install.packages("devtools")
library(devtools)
remotes::install_github("agrakhi/GDSARM")
```

### Bug reports

Please submit any bugs or issues (or suggestions) using the [issues](https://github.com/agrakhi/GDSARM/issues) tab of the repo.

## Usage

The main functions users will use are `GDSARM` and `GDS_givencols`. `GDSARM` is the main method. `GDS_givencols` is the helper function which helps user to run GDS on only main effects and on main effects plus all two-factor interactions.

```{r}
library(GDSARM)

data(dataHamadaWu)
X = dataHamadaWu[,-8]
Y= dataHamadaWu[,8]
delta.n=10
n = dim(X)[1]
m = dim(X)[2]
nint = ceiling(0.2*choose(m,2))
nrep = choose(m,2)
ntop = max(20, nint*nrep/(2*choose(m,2)))
pkeep = 0.25 
cri.penter = 0.01
cri.premove = 0.05
design = X
# GDS-ARM with default values
GDSARM(delta.n,nint, nrep, ntop,pkeep, X, Y,cri.penter, cri.premove)

# GDS-ARM with default values but with weak heredity
opt.heredity="weak" 
GDSARM(delta.n,nint, nrep, ntop,pkeep, X,Y,cri.penter, cri.premove,opt.heredity)


data(dataHamadaWu)
X = dataHamadaWu[,-8]
Y= dataHamadaWu[,8]
delta.n=10
# GDS on main effects 
GDS_givencols(delta.n,design = X, Y=Y, which.cols = "main")

# GDS on main effects and two-factor interactions
GDS_givencols(delta.n,design = X, Y=Y)
```

Check out the vignettes for more examples and details.

## License

This package is released in the public domain under the General Public License [GPL](https://www.gnu.org/licenses/gpl-3.0.en.html). 

## References

Singh, R. and Stufken, J. (2022). Factor selection in screening experiments by aggregation over random models, 1--31. https://arxiv.org/abs/2205.13497


NEWS.md
# GDSARM 0.1.0

* Initial release


R/GDS_givencols.R
#' Gauss-Dantzig Selector
#'
#' @description This function runs Gauss-Dantzig Selector on the given columns. 
#' We have two options: either (a) GDS(m) on the \code{m} main 
#' effects, and (b) GDS(m+2fi) on the \code{m} main effects plus two-factor interactions. 
#' For a given \code{delta}, DS minimizes the l1-norm (sum of absolute values)
#' of \code{beta} subject to the constraint that \code{max(t(X)(y-X * beta))} <= \code{delta}.
#' GDS is run for multiple values of \code{delta}. We use kmeans with 2 
#' clusters and BIC to select a best model  for all the \code{delta} values.
#' 
#' @param delta.n a positive integer suggesting the number of deltas
#' to be tried. \code{delta.n} values of \code{delta} will be used 
#' strictly between 0 and \code{max(t(X)y)}. The default value is
#' set to 10.
#' 
#' @param design a \eqn{n \times m}{n x m} matrix of \code{m} two-level main effects. 
#' The columns should have the high-level coded as +1 and the low-level codes as -1.
#' 
#' @param Y a vector of \code{n} responses.
#' 
#' @param which.cols a string with either `main` or `main2fi`. Denotes
#' whether the Gauss-Dantzig Selector should be run on the main effect columns (`main`), or on  
#' all main effects plus all 2 factor interaction columns (`main2fi`). 
#' The default value is `main2fi`.
#' 
#' @return A list returning the effects identified as active as well as the
#' corresponding identified important factors.
#' 
#' @source Cand{\`e}s, E. and Tao, T. (2007). The Dantzig selector: Statistical estimation when p is much
#' larger than n. Annals of Statistics 35 (6), 2313--2351.
#' 
#' @source Dopico-Garc{\' i}a, M.S., Valentao, P., Guerra, L., Andrade, P. B., and Seabra, R. M. (2007).
#' Experimental design for extraction and quantification of phenolic 
#' compounds and organic acids in white "Vinho Verde" grapes 
#' Analytica chimica acta, 583(1): 15--22.
#' 
#' @source Hamada, M. and Wu, C. F. J. (1992). Analysis of designed experiments 
#' with complex aliasing. Journal of Quality Technology 24 (3), 130--137.
#' 
#' @source Hunter, G. B., Hodi, F. S. and Eagar, T. W. (1982). High cycle fatigue of weld repaired
#' cast Ti-6AI-4V. Metallurgical Transactions A 13 (9), 1589--1594.
#' 
#' @source Phoa, F. K., Pan, Y. H. and Xu, H. (2009). Analysis of supersaturated 
#' designs via the Dantzig selector. Journal of Statistical Planning and Inference
#' 139 (7), 2362--2372.
#'
#' @source Singh, R. and Stufken, J. (2022). Factor selection in screening experiments
#' by aggregation over random models, 1--31. \url{https://arxiv.org/abs/2205.13497/}
#' 
#' @seealso \code{\link{GDSARM}}, \code{\link{dantzig.delta}}
#'
#' @examples
#' data(dataHamadaWu)
#' X = dataHamadaWu[,-8]
#' Y= dataHamadaWu[,8]
#' delta.n=10
#' # GDS on main effects 
#' GDS_givencols(delta.n,design = X, Y=Y, which.cols = "main")
#' 
#' # GDS on main effects and two-factor interactions
#' GDS_givencols(delta.n,design = X, Y=Y)
#' 
#' data(dataCompoundExt)
#' X = dataCompoundExt[,-9]
#' Y= dataCompoundExt[,9]
#' delta.n=10
#' # GDS on main effects
#' GDS_givencols(delta.n,design = X, Y=Y, which.cols = "main")
#' # GDS on main effects and two-factor interactions
#' GDS_givencols(delta.n,design = X, Y=Y, which.cols = "main2fi")
#' @export
#'

GDS_givencols<- function(delta.n=10, design, Y, which.cols = c("main2fi")){
  n = dim(design)[1]
  m = dim(design)[2]
  ncluster = 2
  artificial.response <- stats::rnorm(n,0,1)
  
  Xall <- as.data.frame(stats::model.matrix(stats::lm(artificial.response ~ .^2, design)))[,-1]
  Xall.scale = as.data.frame(base::scale(Xall,center = T,scale = T)) #*(1/sqrt(n-1))*(0.2)^12
  
  if (which.cols == "main") {
    Xran = Xall.scale[,1:m]
  } else {
    Xran = Xall.scale
  }

  ncol = ncol(Xran)
  delta0 <- max(abs(t(Xran)%*%Y))
  delta = seq(0, delta0, length.out = delta.n+2)
  delta = delta[-1]
  
  beta = dantzig.delta(Xran, Y, delta, plot=F)
  beta2 = beta;
  beta2[]=0
  
  BICafterkmeans = numeric(delta.n-1)
  BICafterkmeans = numeric(delta.n-1)
  for (ct_lam in 1:(delta.n-1)){
    ## doing kmeans
    tempvec = beta[ct_lam,]
    tempvec1 = c(tempvec,rep(0,ncol-length(tempvec)))
    if (tempvec1[order(tempvec1, decreasing = T)][ncluster-1]> 0) {
      kmeansOUT=stats::kmeans(abs(tempvec1), ncluster) #$cluster
      clusternumber = order(kmeansOUT$centers)[ncluster] #pick the cluster
      #with largest cluster mean
      selectcols = tempvec[which(kmeansOUT$cluster==clusternumber)]
      # print(selectcols)
      beta2[ct_lam,names(selectcols)]=1
      beta2[ct_lam,names(selectcols)]=1
      datasubset = as.data.frame(cbind(Xran[,names(selectcols),drop=F],Y))
      colnames(datasubset) = c(names(selectcols),"Y")
      ## fit a model to find BIC
      fit = stats::lm(Y~., data=datasubset)
      BICafterkmeans[ct_lam] =stats::AIC(fit, k = log(n))#BIC
    } else {
      BICafterkmeans[ct_lam] = 10^8
    }
  }
  Index_minBIC = which.min(BICafterkmeans)
  SelectedEffects = colnames(beta2[Index_minBIC,which(beta2[Index_minBIC,]==1),drop=F])
  
  SelectedFactors1 = strsplit(SelectedEffects,":")
  SelectedFactors = unique(c(unlist(SelectedFactors1)))
  
  return(list(effects = SelectedEffects, factors =SelectedFactors))
}


R/dantzig.delta.R
#' Dantzig selector with an option to make profile plot
#'
#' @description The Dantzig selector (DS) finds a solution for the model parameters
#' of a linear model, \code{beta} using linear programming. For a given \code{delta},
#' DS minimizes the l1-norm (sum of absolute values) of \code{beta} subject to the constraint
#' that \code{max(t(X)(y-X * beta)) <= delta}.
#'
#' @param X a design matrix.
#'
#' @param y a vector of  responses.
#'
#' @param delta the specific value of \code{delta} for which the DS
#' optimization needs to be solved
#'
#' @param plot a boolean value of either TRUE or FALSE with TRUE
#' indicating that the profile plot should be drawn.
#'
#' @return A matrix of the estimated values of \code{beta} with each 
#' row corresponding to a particular value of \code{delta}.
#'
#' @seealso \code{\link{GDS_givencols}}, \code{\link{GDSARM}}
#'
#' @source Cand{\`e}s, E. and Tao, T. (2007). The Dantzig selector: Statistical estimation when p is much
#' larger than n. Annals of Statistics 35 (6), 2313--2351.
#' 
#' @source Phoa, F. K., Pan, Y. H. and Xu, H. (2009). Analysis of supersaturated 
#' designs via the Dantzig selector. Journal of Statistical Planning and Inference
#' 139 (7), 2362--2372.
#' 
#' @examples
#' data(dataHamadaWu)
#' X = dataHamadaWu[,-8]
#' Y= dataHamadaWu[,8]
#' #scale and center X and y
#' scaleX = base::scale(X, center= TRUE, scale = TRUE)
#' scaleY = base::scale(Y, center= TRUE, scale = FALSE)
#' maxDelta = max(t(scaleX)%*%matrix(scaleY, ncol=1))
#' # Dantzig Selector on 4 delta values equi spaced between 0 and maxDelta
#' dantzig.delta(scaleX, scaleY, delta = seq(0,maxDelta,length.out=4)) 
#'
#' @export

dantzig.delta = function(X, y, delta, plot=F)
{
  beta = NULL
  for(i in 1:length(delta)) beta = rbind(beta, dantzigS(X, y, delta[i])$beta)
  dimnames(beta)[[1]]=delta
  if(plot) graphics::matplot(delta, beta, type="l")  # profile plot
  beta
}


R/StepI_chooseints.R
#' Step I: Multiple GDS runs with random interactions
#'
#' @description Runs Gauss Dantzig Selector (GDS) multiple times, each time 
#' with a differently set of randomly selected two-factor interactions.
#' All \code{m} main effects are included in each GDS run. For each set of
#' randomly selected interactions, the best GDS output is chosen among 
#' \code{delta.n} values of \code{delta}. We use kmeans with 2 
#' clusters and BIC to select such best model.
#'
#'
#' @param delta.n a positive integer suggesting the number of delta values
#' to be tried. \code{delta.n} values of \code{delta} will be used 
#' strictly between 0 and \code{max(t(X)y)}. The default value is set to 10.
#' 
#' @param nint a positive integer representing the number of randomly 
#' chosen interactions. The suggested value to use is the ceiling of 20% 
#' of the total number of interactions, that is, for \code{m} factors, we have
#' \code{ceiling(0.2(m choose 2))}.
#' 
#' @param nrep a positive integer representing the number of times GDS should
#' be run. The suggested value is \code{(m choose 2)}.
#' 
#' @param Xmain a \eqn{n \times m}{n x m} matrix of \code{m} main effects.
#' 
#' @param Xint a matrix of \eqn{{m \choose 2}}{\code{m choose 2})} two-factor 
#' interactions.
#' 
#' @param Y a vector of \code{n} responses.
#' 
#' @param opt.heredity a string with either `none`, or `weak`, or `strong`. Denotes
#' whether the effect-heredity (weak or strong) should be embedded in GDS-ARM. 
#' The default value is `none` as suggested in Singh and Stufken (2022).
#' 
#' @return A list containing the (a) matrix of the output of each GDS run with each 
#' row representing the selected effects from the corresponding GDS run, (b) 
#' a vector with the corresponding BIC values of each model.
#'
#' @source Singh, R. and Stufken, J. (2022). Factor selection in screening experiments
#' by aggregation over random models, 1--31. \url{https://arxiv.org/abs/2205.13497/}
#' 
#' @keywords internal
#'
## needs delta.n, nint, nrep,design as Xmain (main effects), Xint (two factor interactions)
StepI_chooseints <- function(delta.n=10,nint,nrep,Xmain,Xint,Y, opt.heredity = c("none")) {
  n = dim(Xmain)[1]
  m= dim(Xmain)[2]
  
  ncluster = 2
  
  ncol = dim(Xmain)[2]+dim(Xint)[2]
  ListofSelectedEffects=c()
  all.nint=matrix(ncol = nint,nrow = nrep)
  for (i in 1:nrep) {
    all.nint[i,]=sample((dim(Xmain)[2]+1):ncol,nint,replace = F)
  }
  B=matrix(ncol = ncol,nrow = nrep)
  BICReps= numeric(nrep)
  res = matrix(0,ncol = ncol,nrow = nrep)
  for (irep in 1:nrep) {
    Xran=cbind(Xmain,Xint[,all.nint[irep,]-dim(Xmain)[2]])

    delta0 <- max(abs(t(Xran)%*%Y))
    delta = seq(0, delta0, length.out = delta.n+2)
    delta = delta[-1]

    beta = dantzig.delta(Xran, Y, delta, plot=F)
    beta2 = beta;
    beta2[]=0

    #matrix(0,nrow=nrow(beta),ncol=ncol(beta))
    BICafterkmeans = numeric(delta.n-1)
    BICafterkmeans = numeric(delta.n-1)
    for (ct_lam in 1:(delta.n-1)){
      ## doing kmeans
      tempvec = beta[ct_lam,]
      tempvec1 = c(tempvec,rep(0,ncol-length(tempvec)))
      if (tempvec1[order(tempvec1, decreasing = T)][ncluster-1]> 0) {
        kmeansOUT=stats::kmeans(abs(tempvec1), ncluster) #$cluster
        clusternumber = order(kmeansOUT$centers)[ncluster] #pick the cluster
        #with largest cluster mean
        selectcols = tempvec[which(kmeansOUT$cluster==clusternumber)]
        # print(selectcols)
        beta2[ct_lam,names(selectcols)]=1
        beta2[ct_lam,names(selectcols)]=1
		#heredity or not
    namesselectcols = names(selectcols)
   # namesselectcols = (selectcols)
    namesselectints = grep(":", namesselectcols, value = F)
    deleteInt = numeric(length(namesselectints))
    deleteInt[] = 1
    namesselectmain = namesselectcols[setdiff(1:length(namesselectcols),namesselectints)]
   if ((length(namesselectints)) >= 1) {
   for (tt in 1:length(namesselectints)){
    selectint = namesselectcols[namesselectints[tt]]
    selectint1 = unlist(strsplit(selectint,":"))
    if (opt.heredity == "weak" ) {
      if (selectint1[1] %in% namesselectmain | selectint1[2] %in% namesselectmain){
      deleteInt[tt] = 0
      }
    } else if (opt.heredity == "strong" ) {
      if (selectint1[1] %in% namesselectmain & selectint1[2] %in% namesselectmain){
        deleteInt[tt] = 0
      }
    } else {
	 deleteInt[tt] = 0
	}
  }
  namesToUse = c(namesselectmain,namesselectcols[namesselectints[which(deleteInt == 0)]])
  } else {
    namesToUse = namesselectmain  
  }
        datasubset = as.data.frame(cbind(Xran[,namesToUse,drop=F],Y))
        colnames(datasubset) = c(namesToUse,"Y")
		## fit a model to find BIC
        fit = stats::lm(Y~., data=datasubset)
        BICafterkmeans[ct_lam] =stats::AIC(fit, k = log(n))#BIC
      } else {
        BICafterkmeans[ct_lam] = 10^8
      }
    }
    Index_minBIC = which.min(BICafterkmeans)
    SelectedEffectsFromThisRep = colnames(beta2[Index_minBIC,which(beta2[Index_minBIC,]==1),drop=F])
    if (length(SelectedEffectsFromThisRep)> 0) {
      B[irep,1:length(SelectedEffectsFromThisRep)]=SelectedEffectsFromThisRep
    }
    BICReps[irep]=BICafterkmeans[Index_minBIC]
  }

  return(list(output=B,BIC=BICReps))
}



GDSARM/R/StepIII_stepwise.R
#' Step III: Stepwise on the consolidated output from different GDS runs
#'
#' @description Runs the stepwise regression on the output received from 
#' top models of the consolidated output of different GDS runs. With
#' \code{n} being the number of runs, the stepwise regression starts
#' with at most \code{(n-3)} selected effects from the previous step. The 
#' remaining effects from the previous step as well as all main effects are
#' given a chance to enter into the model using the forward-backward stepwise
#' regression.
#'
#' @param xstart a vector with effects' names corresponding to the starting model.
#'  
#' @param xremain a vector with effects' names corresponding to the remaining 
#' main effects and other effects that needs to be explored with stepwise regression.
#'  
#' 
#' @param Xmain a \eqn{n \times m}{n x m} matrix of \code{m} main effects.
#' 
#' @param Xint a matrix of \code{m choose 2} two-factor 
#' interactions.
#' 
#' @param Y a vector of \code{n} responses.
#' 
#' @param cri.penter the p-value cutoff for an effect to enter into the 
#' stepwise regression model
#' 
#' @param cri.premove the p-value cutoff for an effect to exit from the 
#' stepwise regression model
#'
#' @param opt.heredity a string with either `none`, or `weak`, or `strong`. Denotes
#' whether the effect-heredity (weak or strong) should be embedded in GDS-ARM. 
#' The default value is `none` as suggested in Singh and Stufken (2022).
#' 
#' @return A list returning the effects identified as active as well as the
#' corresponding identified important factors.
#'
#' @source Singh, R. and Stufken, J. (2022). Factor selection in screening experiments
#' by aggregation over random models, 1--31. \url{https://arxiv.org/abs/2205.13497/}
#' 
#' @keywords internal
#'
StepIII_stepwise<-function(xstart,xremain,Xmain, Xint, Y,cri.penter=0.01, cri.premove=0.05,opt.heredity="none"){

  nrow = dim(Xmain)[1]
  Xall = cbind(Xmain, Xint)
  
  #dataset needed for stepwise only
  data=as.data.frame(cbind(Xall,Y))
  colnames(data)[dim(data)[2]]="y"
  colnames(data)=gsub(":", "a",colnames(data))
  
  ## 2nd selection -- backward
  modelfin=0
  deadloop=0

  while(length(xstart)+2<nrow & modelfin==0){

    deadloop=deadloop+1
    if(deadloop>30){break}
    movefor=1
    moveback=1

    #forward
    model = paste(xstart,sep = "",collapse ="+")
    if(model==""){model="y~1"}else{
      model = paste("y~1+",model,sep = "")}


    if (length(xremain)>= 1) {
      criteria=vector("numeric",length(xremain))
      for (i in 1:length(xremain)) {
        ##print("hi")
        model2=paste(model,"+",xremain[i],sep = "")
        temp.model2 <- stats::lm(model2, data = data)
        pv=summary(temp.model2)$coefficients[-1,4]
        criteria[i]=pv[length(pv)]
      }
      deleposi=which(criteria==min(criteria)) #r2--max
      if(min(criteria)>cri.penter){movefor=0}else{
        xstart=c(xstart,xremain[deleposi])
        xremain=xremain[-deleposi]
      }
    } else{movefor=0}
    #backward
    model = paste(xstart,sep = "",collapse ="+")
    if(model==""){
      model="y~1"
      max.pvalue=0
    }else{
      model = paste("y~1+",model,sep = "")
      temp.model <- stats::lm(model, data = data)
      pvalue <- base::summary(temp.model)$coefficients[-1,4]
      max.pvalue=max(pvalue)}
    if(max.pvalue<=1 & max.pvalue>cri.premove){
      deleposi = which(xstart==names(pvalue)[which(pvalue==max.pvalue)])
      xremain=c(xremain,xstart[deleposi])
      xstart=xstart[-deleposi]
    }else{moveback=0}

    if(length(xstart)==0){modelfin=1}
    if(length(xremain)==0){modelfin=1}
    if(length(xstart)>=nrow){modelfin=1}
    if(moveback==0 & movefor==0){modelfin=1}
  }

  if(length(xstart)==0){
    result=character(0)
  }else{
    result=matrix(nrow=length(xstart),ncol=2)
    for(i in 1:length(xstart)){
      result[i,]= unlist(strsplit(xstart[i],split="a"))
      if(result[i,1]==result[i,2]){result[i,2]=""}
    }
    result = as.matrix(result)}

  xstarta=gsub("a", ":",xstart)
  xstarta = unique(xstarta)
  #embed heredity or not
  namesselectcols = xstarta
  namesselectints = grep(":", namesselectcols, value = F)
  deleteInt = numeric(length(namesselectints))
  deleteInt[] = 1
  namesselectmain = namesselectcols[setdiff(1:length(namesselectcols),namesselectints)]
  if ((length(namesselectints)) >= 1) {
    for (tt in 1:length(namesselectints)){
      selectint = namesselectcols[namesselectints[tt]]
      selectint1 = unlist(strsplit(selectint,":"))
      if (opt.heredity == "weak" ) {
        if (selectint1[1] %in% namesselectmain | selectint1[2] %in% namesselectmain){
          deleteInt[tt] = 0
        }
      } else if (opt.heredity == "strong" ) {
        if (selectint1[1] %in% namesselectmain & selectint1[2] %in% namesselectmain){
          deleteInt[tt] = 0
        }
      } else {
	 deleteInt[tt] = 0
	}
    }
    namesToUse = c(namesselectmain,namesselectcols[namesselectints[which(deleteInt == 0)]])
  } else {
    namesToUse = namesselectmain  
  }
  xstarta = unique(namesToUse)
  result1 = unique(unlist(strsplit(xstarta,":")))
  return(list(effects=xstarta, factors=result1) )
}


GDSARM/R/GDSARM.R
#' Gauss-Dantzig Selector - Aggregation over Random Models (GDS-ARM)
#'
#' @description The GDS-ARM procedure consists of three steps. First, it runs 
#' runs Gauss Dantzig Selector (GDS) \code{nrep} times, each time 
#' with a differently set of \code{nint} randomly selected two-factor interactions.
#' All \code{m} main effects are included in each GDS run. Second, the best
#' \code{ntop} models are identified with the smallest BIC. The effects that are 
#' declared as active effects in at least \code{pkeep x ntop} of the \code{ntop} 
#' models are then passed on to the third stage. In the third stage, these effects are
#' passed through the stepwise regression model. With
#' \code{n} being the number of runs, the stepwise regression starts
#' with at most \code{n-3} selected effects from the previous step. The 
#' remaining effects from the previous step as well as all main effects are
#' given a chance to enter into the model using the forward-backward stepwise
#' regression. The function also has the option of using the modified GDS-ARM. 
#' The modified version incorporates effect heredity in two steps, first, for 
#' each model found by GDS, we ignore active interactions when at least one of 
#' the main effects is not active (for weak heredity) or when both main effects are not
#' active (for strong heredity); and second, we do the same for the model found 
#' after the stepwise stage of GDS-ARM.
#' 
#' @param delta.n a positive integer suggesting the number of delta values
#' to be tried. \code{delta.n} values of \code{delta} will be used 
#' strictly between 0 and \code{max(t(X)y)}. The default value is set to 10.
#' 
#' @param nint a positive integer representing the number of randomly 
#' chosen interactions. The suggested value to use is the ceiling of 20% 
#' of the total number of interactions, that is, for \code{m} factors, we have
#' \code{ceiling(0.2(m choose 2))}.
#' 
#' @param nint a positive integer representing the number of randomly 
#' chosen interactions. The suggested value to use is the ceiling of 20% 
#' of the total number of interactions, that is, for \code{m} factors, we have
#' \code{ceiling(0.2(m choose 2))}.
#' 
#' @param nrep a positive integer representing the number of times GDS should
#' be run. The suggested value is \code{(m choose 2)}.
#' 
#' @param ntop a positive integer representing the number of top models 
#' to be selected among the \code{nrep} models. The suggested value is 
#' \code{max(20, (nrep x nint)/(m(m-1))}.
#' 
#' @param pkeep a number between 0 and 1 representing the proportion of \code{ntop}
#' models in which an effect needs to appear in order to be declared active.
#' 
#' @param design a \eqn{n \times m}{n x m} matrix of \code{m} two-level main effects. 
#' The columns should have the high-level coded as +1 and the low-level codes as -1.
#' 
#' @param Y a vector of \code{n} responses.
#' 
#' @param cri.penter the p-value cutoff for an effect to enter into the 
#' stepwise regression model. The suggested value is 0.01.
#' 
#' @param cri.premove the p-value cutoff for an effect to exit from the 
#' stepwise regression model. The suggested value is 0.05.
#' 
#' @param opt.heredity a string with either `none`, or `weak`, or `strong`. Denotes
#' whether the effect-heredity (weak or strong) should be embedded in GDS-ARM. 
#' The default value is `none` as suggested in Singh and Stufken (2022).
#' 
#' @return A list returning the effects identified as active as well as the
#' corresponding identified important factors.
#' 
#' @source Cand{\`e}s, E. and Tao, T. (2007). The Dantzig selector: Statistical estimation when p is much
#' larger than n. Annals of Statistics 35 (6), 2313--2351.
#' 
#' @source Dopico-Garc{\' i}a, M.S., Valentao, P., Guerra, L., Andrade, P. B., and Seabra, R. M. (2007).
#' Experimental design for extraction and quantification of phenolic 
#' compounds and organic acids in white "Vinho Verde" grapes 
#' Analytica chimica acta, 583(1): 15--22.
#' 
#' @source Hamada, M. and Wu, C. F. J. (1992). Analysis of designed experiments 
#' with complex aliasing. Journal of Quality Technology 24 (3), 130--137.
#' 
#' @source Hunter, G. B., Hodi, F. S. and Eagar, T. W. (1982). High cycle fatigue of weld repaired
#' cast Ti-6AI-4V. Metallurgical Transactions A 13 (9), 1589--1594.
#'
#' @source Phoa, F. K., Pan, Y. H. and Xu, H. (2009). Analysis of supersaturated 
#' designs via the Dantzig selector. Journal of Statistical Planning and Inference
#' 139 (7), 2362--2372.
#' 
#' @source Singh, R. and Stufken, J. (2022). Factor selection in screening experiments
#' by aggregation over random models, 1--31. \url{https://arxiv.org/abs/2205.13497/}
#' 
#' @seealso \code{\link{GDS_givencols}}, \code{\link{dantzig.delta}}
#'
#' @examples
#' data(dataHamadaWu)
#' X = dataHamadaWu[,-8]
#' Y= dataHamadaWu[,8]
#' delta.n=10
#' n = dim(X)[1]
#' m = dim(X)[2]
#' nint = ceiling(0.2*choose(m,2))
#' nrep = choose(m,2)
#' ntop = max(20, nint*nrep/(2*choose(m,2)))
#' pkeep = 0.25 
#' cri.penter = 0.01
#' cri.premove = 0.05
#' design = X
#' # GDS-ARM with default values
#' GDSARM(delta.n,nint, nrep, ntop,pkeep, X, Y,cri.penter, cri.premove)
#' 
#' # GDS-ARM with default values but with weak heredity
#' opt.heredity="weak" 
#' GDSARM(delta.n,nint, nrep, ntop,pkeep, X,Y,cri.penter, cri.premove,opt.heredity)
#' 
#' 
#' data(dataCompoundExt)
#' X = dataCompoundExt[,-9]
#' Y= dataCompoundExt[,9]
#' delta.n=10
#' n = dim(X)[1]
#' m = dim(X)[2]
#' nint = ceiling(0.2*choose(m,2))
#' nrep = choose(m,2)
#' ntop = max(20, nint*nrep/(2*choose(m,2)))
#' pkeep = 0.25 
#' cri.penter = 0.01
#' cri.premove = 0.05
#' design = X
#' # GDS-ARM on compound extraction
#' GDSARM(delta.n,nint, nrep, ntop,pkeep, X,Y,cri.penter, cri.premove)
#' 
#' # GDS-ARM on compound extraction with strong heredity
#' opt.heredity="strong"
#' GDSARM(delta.n,nint, nrep, ntop,pkeep, X,Y,cri.penter, cri.premove,opt.heredity)
#' 
#' 
#' @export
#'

GDSARM<- function(delta.n=10, nint, nrep, ntop,pkeep, design, Y,cri.penter=0.01, cri.premove=0.05,opt.heredity=c("none")){
  n = dim(design)[1]
  m = dim(design)[2]
  artificial.response <- stats::rnorm(n,0,1)

  set.seed(1234) 
  
  Xall <- as.data.frame(stats::model.matrix(stats::lm(artificial.response ~ .^2, design)))[,-1]
  Xall.scale = as.data.frame(base::scale(Xall,center = T,scale = T)) #*(1/sqrt(n-1))*(0.2)^12
  Xmain = Xall.scale[, c(1:m)]
  Xint = Xall.scale[,-c(1:m) ]

        # step I: run many GDS models
        StepIResults = StepI_chooseints(delta.n,nint,nrep,Xmain, Xint,Y,opt.heredity)

        #selecting the top ntop models
        TopNtopModels=base::order(StepIResults$BIC)[1:ntop]
        SelectedEffectsFromTopnNmodels = StepIResults$output[TopNtopModels,]

        # counting frequencies as per columns
        FreqSelectedEffects=base::table(SelectedEffectsFromTopnNmodels)
        FreqSelectedEffects = base::sort(FreqSelectedEffects, decreasing = T)
        select = names(FreqSelectedEffects[which(FreqSelectedEffects>=pkeep*ntop)])

        # step III: stepwise
         if (length(select) < (n-2)){
            xstartmodel=gsub(":", "a",select)
            xaddall = unique(c(xstartmodel,colnames(Xmain)))
            xremain = xaddall[-(1:length(xstartmodel))]
            xaddall = unique(xaddall)
            xremain = unique(xremain)
          } else{
            xstartmodel=gsub(":", "a",select)[1:(n-3)]
            xaddall = unique(c(xstartmodel,gsub(":", "a",select)[-(1:(n-3))],colnames(Xmain)))
            xremain = xaddall[-(1:length(xstartmodel))]
            xaddall = unique(xaddall)
            xremain = unique(xremain)
          }
          OutputStepwise = StepIII_stepwise(xstartmodel,xremain,Xmain, Xint, Y,cri.penter, cri.premove,opt.heredity)
    return(OutputStepwise)
}
